### DISCUSSION TITLE
Update & Delete MongoDB

### DISCUSSION THREAD

**Discussion by kebab-case-dev** [neuefische/questions#548] [Web] [answered]
My Form for updating and deleting mongoDB entries works not as i want

My [Pull Request](https://github.com/bk1405bgl/capstone-project/pull/19)

Code Preview:

```
import Link from "next/link";
import Image from "next/image";
import React, { useState } from "react";
import useSWRMutation from "swr/mutation";

export default function Detail({ comic, friends }) {
  // const [friendId, setFriendId] = useState({});
  // if (!comic || !friends) return <h3>Loading</h3>;
  const { id, title, upc, description, images } = comic;
  const { trigger, isMutating } = useSWRMutation(
    `/api/comics/${title}`,
    sendRequest
  );
  async function sendRequest(url, { arg }) {
    const response = await fetch(url, {
      method: "PUT",
      body: JSON.stringify(arg),
      headers: {
        "Content-Type": "application/json",
      },
    });
    if (response.ok) {
      await response.json();
    } else {
      console.error(`Error: ${response.status}`);
    }
  }
  async function handleRent(event) {
    event.preventDefault();
    const formData = new FormData(event.target);
    const comicData = Object.fromEntries(formData);
    await trigger(comicData);
  }
  async function handleDelete() {
    await fetch(`/api/comics/${comic._id}`, {
      method: "DELETE",
    });
  }
  async function handleNoFriend(event) {
    event.preventDefault();
  }
  return (
    <main>
      <Link href="/comics">Zurück zur Übersicht</Link>
      <div>
        <div>
          <form onSubmit={handleRent}>
            <select name="friends">
              <option value="">Freund auswählen</option>
              {friends?.map((friend) => {
                return (
                  <option key={friend._id} value={friend._id}>
                    {friend.friendName}
                  </option>
                );
              })}
            </select>
            <button type="submit">Ausleihen</button>
          </form>
          <form>
            <button type="button" onClick={() => handleDelete()}>
              Aus Sammlung entfernen
            </button>
            <button type="button" onClick={handleNoFriend}>
              Vom Freund zurück erhalten
            </button>
          </form>
          <Image
            src={`${images[0].path}/portrait_xlarge.${images[0].extension}`}
            alt={title}
            width={150}
            height={225}
            priority
          />
        </div>
        <h4>Titel: {title}</h4>
        <p>
          <b>EAN/UPC/ISBN:</b> {upc}
        </p>
        <p>
          <b>Beschreibung:</b> {description}
        </p>
      </div>
    </main>
  );
}
```

**Comment by PReichetanz** [comment, marked as answer]
In the corresponding [API route](https://github.com/bk1405bgl/capstone-project/blob/dpoc/pages/api/comics/%5Bid%5D.js), you use `.findByIdAndUpdate(title, ...)` and `.findByIdAndDelete(title)`.

As far as I can see, `title` is not an `id` which `mongoose` can use to find the entry, but is rather an entire object. You could check whether you pass the correct `id` to this method.

**Reply by kebab-case-dev** [reply]
Thank you so much. I changed this to id. The delete button works, except the push("/comics"); Line throws an error push is not defined.

And i still do have problems with updating a record.

**Reply by PReichetanz** [reply]
> Thank you so much. I changed this to id. The delete button works, except the push("/comics"); Line throws an error push is not defined.
> 
> And i still do have problems with updating a record.

I don't see any `push("/comics")` in the code you've posted; please provide more information about the errors and when they occur.

**Reply by kebab-case-dev** [reply]
i tried with:

```
async function handleDelete() {
    await fetch(`/api/comics/${comic._id}`, {
      method: "DELETE",
    });
    router.back();
  }
```

and with:

```
async function handleDelete() {
    await fetch(`/api/comics/${comic._id}`, {
      method: "DELETE",
    });
    push("/comics");
  }
```

**Reply by kebab-case-dev** [reply]
i solved the push problem with routing. know i have just problems with updating data in db

**Reply by PReichetanz** [reply]
Nice that the routing works now!

Regarding updating the db: Please give us some information what exactly is the issue. What does work, what does not work, and what error messages do you have.

**Reply by kebab-case-dev** [reply]
> Nice that the routing works now!
> 
> Regarding updating the db: Please give us some information what exactly is the issue. What does work, what does not work, and what error messages do you have.

Hi Paul,

i try to fetch from two collections in one file and push one information from one collection into the other one.
it means, i try to put the friend id into a commic object. but it could be, that this is impossible. actually, i receive this error message: 

`error - CastError: Cast to ObjectId failed for value "Hedge Knight II: Sworn Sword (2007) " (type string) at path "_id" for model "Comic" ...`

i am still trying to impelement the solution from the handouts.

would also be nice, if we could meet in a zoom room so i can show you my project better.

**Reply by PReichetanz** [reply]
Please commit your latest try and give us a link here. I guess the error has to do with your Schema, but I cannot say anything specific before looking into the code.

**Reply by kebab-case-dev** [reply]
> Please commit your latest try and give us a link here. I guess the error has to do with your Schema, but I cannot say anything specific before looking into the code.

My working tree is clean and the link is the same.

[PR](https://github.com/bk1405bgl/capstone-project/pull/19)

**Reply by PReichetanz** [reply]
Hi @bk1405bgl, I had a look into your code and found a solution:

```js
// components/DatailsPageOwn/index.js
export default function Detail({ comic, friends }) {
...
// use the `_id` for your API route call instead of `title` (this is why you've gotten the error mentioned above
  const { trigger, isMutating } = useSWRMutation(
    `/api/comics/${_id}`,
    sendRequest
  );
// in the return statement, adjust the `name` value to your Comic schema where it is called `friendId` instead of `friends`
<form onSubmit={handleRent}>
            <select name="friendId">

}
```

Finally, I guess you don't want to save the ObectId of another document as a simple string, so you have to adjust your schema:

```js
// models/Comic.js
const comicSchema = new Schema({
...
friendId: { type: Schema.Types.ObjectId },
})

```

By doing this, I am able to select a friend and save its `ObjectId` to the collection of the the comic. Hope that helps 🚀

**Reply by kebab-case-dev** [reply]
Thank you @PReichetanz, this works well.
to finish my functionality, i need to unset the friendId in the comic object, or delete the field would be ok, too. but i cannot find anything in our handouts. the web shows mongodb commands to unset a field, but i want to do it with a button click...

working tree is clean

[PR](https://github.com/bk1405bgl/capstone-project/pull/19)

**Reply by PReichetanz** [reply]
Please follow the guide how to ask good questions again, because without any further information, I cannot do anything:
Where exactly have you tried to insert the logic? Which button do you want to click to unset the field? What have you tried so far?

The link to the PR does not answer these questions.

**Reply by kebab-case-dev** [reply]
I added a function handleNoFriend:

```
async function handleNoFriend(event) {
    event.preventDefault();
    const formData = new FormData(event.target);
    const comicData = Object.fromEntries(formData);
    await trigger(comicData);
  }
```

i made a new form with onSubmit={handleNoFriend}:

```
<form onSubmit={handleNoFriend}>
            <button type="submit">
              Vom Freund zurück erhalten
            </button>
          </form>
```

i tried to set a value in the dropdown menu with "In eigener Sammlung" with a value=0 and value=$unset but it did not work. this would not be good for the usability.

I tried to insert the functionality in the same component liek the functions before, where you helped me out. It is the component: `DetailsOwnPage/index.js`

I hope this is a good question

**Reply by PReichetanz** [reply]
The problem is twofold: you need a second `PUT` route, but cannot place it in the same API route; and you don't have (nor need) a form to handle the update.

My solution would be to create a new API route called `/api/comics/friend/[id].js`, use the `PUT` method and the mongoose method`$unset`:

```js
import dbConnect from '../../../../db/connect';
import Comic from '../../../../db/models/Comic';

export default async function handler(request, response) {
  await dbConnect();
  const { id } = request.query;

  if (request.method === 'PUT') {
    const comicToUpdate = await Comic.findByIdAndUpdate(id, {
      $unset: { friendId: '' },
    });
    response.status(200).json(comicToUpdate);
  }
}
```

You can now call this API route in your 'handleNoFriend` function in `components/DetailsPageOwn/index.js`:
```js
async function handleNoFriend() {
    await fetch(`/api/comics/update/${comic._id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }
```
Note that there is no need for the `body` key because we are not going to do anything with it (nor do we have information we could pass as `body`).

Finally, pass the `handleNoFriend` function to the button as `onClick` prop:

```jsx
// components/DetailsPageOwn/index.js
<button type="button" onClick={handleNoFriend}>
   Vom Freund zurück erhalten
</button>
```

Cheers ✨

**Reply by kebab-case-dev** [reply]
thank you so much. the only correctin i had to make was in the handleNoFriend function.

you wrote:
```
async function handleNoFriend() {
    await fetch(`/api/comics/update/${comic._id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }
```

i made: 
```
async function handleNoFriend() {
    await fetch(`/api/comics/friend/${comic._id}`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
    });
  }
```
